cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table to be cleared.  
+  
+Example:  
+call cclutClearMockData("person")
+
+**cclutAddMockImplementation(originalName = vc, replaceName = vc)**
+
+Adds a mock implementation to be utilized by cclutExecuteProgramWithMocks.  This can include record structures, subroutines, or other programs.  originalName and replaceName are required.  If originalName is already being mocked, the mock will be overwritten with the new replaceName.  
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object to be mocked.  
+@param replaceName  
+&nbsp;&nbsp;&nbsp;&nbsp;The mocked object.  
+  
+Example:  
+call cclutAddMockImplementation("uar_get_code_by", "mock_uar_get_code_by")
+
+**cclutRemoveMockImplementation(originalName = vc)**
+
+Removes a mock implementation.  
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object that is mocked.  
+  
+Example:  
+call cclutRemoveMockImplementation("uar_get_code_by")
+
+**cclutRemoveAllMockImplementations(null)**
+
+Removes all mock implementations.  
+  
+Example:  
+call cclutRemoveAllMockImplementations(null)
+
+**cclutExecuteProgramWithMocks(programName = vc, params = vc, namespace = vc)**
+
+Executes a program with all mocks currently added through cclutAddMockImplementation() and cclutCreateMockTable(). programName is required.  params is a string parameter to be sent directly to the program, so all commas and string delimiters must be specified.  If namespace is omitted, it will default to the PUBLIC namespace.  
+  
+Example:  
+call cclutExecuteProgramWithMocks("ccl_my_program", "\^MINE^, 1.0, ^string parameter^", "MYNAMESPACE")
+
+**cclutRemoveAllMocks**
+
+Removes all mock implementations and mock tables that have been added through the cclutAddMockImplementation() and cclutCreateMockTable() APIs.  This should be called at the completion of a test suite to clean up all mocks.  
+  
+Example:  
+call cclutRemoveAllMocks(null)
+
+## Implementation Notes
+1. cclutRemoveAllMocks should be called as part of the teardown for all tests.  The framework will attempt to clean up any outstanding mocks, but it is good practice to explicitly remove any mocks to ensure that no mocked tables remain in the Oracle instance.
+
+2. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to children script called from the script-under-test.  Some alternatives would be to mock the child script to return the appropriate data or to mock the child script to execute the real script applying the mocked tables and implementations.
+
+3. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to statements executed through "call parser()" commands.  An alternative would be to mock the parser() call to validate the correct information is supplied, then perform the appropriate mock versions of the actions the statement would normally perform.
```

Comment:
That's fair; however, I'd like to leave the workaround of mocking parser() in the note.  I'm a strong believer in the idea that source code should never have to be modified to fit a testing framework (rather the tests should be able to accommodate any possible source code).  I know we break that a little bit with the tdbexecute workaround, but that's because we have no way within CCL to avoid that.  At least with parser, we do have _a_ way to mock it, so I'd at least like consumers who want to write tests for existing scripts to be aware of an option for testing if they're opposed to modifying their existing source code.